File: api/cutout.js
/**
 * Vercel serverless — same pipeline as python/background_remover.py:
 * BGR → grayscale (OpenCV weights) → binary threshold → BGRA PNG.
 */

const sharp = require("sharp");

const MAX_BYTES = 12 * 1024 * 1024;

function toGray(r, g, b) {
  return Math.round(0.114 * b + 0.587 * g + 0.299 * r);
}

function alphaMask(gray, threshold, invert) {
  if (invert) {
    return gray > threshold ? 0 : 255;
  }
  return gray > threshold ? 255 : 0;
}

async function removeBackground(buffer, threshold, invert) {
  const { data, info } = await sharp(buffer)
    .rotate()
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width, height, channels } = info;
  if (channels < 3) {
    throw new Error("Expected RGB image");
  }

  const out = Buffer.alloc(width * height * 4);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * channels;
      const o = (y * width + x) * 4;
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      const gray = toGray(r, g, b);
      out[o] = r;
      out[o + 1] = g;
      out[o + 2] = b;
      out[o + 3] = alphaMask(gray, threshold, invert);
    }
  }

  return sharp(out, { raw: { width, height, channels: 4 } })
    .png()
    .toBuffer();
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    let size = 0;
    req.on("data", (chunk) => {
      size += chunk.length;
      if (size > MAX_BYTES) {
        reject(new Error("Request body too large"));
        req.destroy();
        return;
      }
      chunks.push(chunk);
    });
    req.on("end", () => resolve(Buffer.concat(chunks)));
    req.on("error", reject);
  });
}

function sendJson(res, status, payload) {
  const body = JSON.stringify(payload);
  res.statusCode = status;
  res.setHeader("Content-Type", "application/json");
  res.setHeader("Access-Control-Allow-Origin", res.getHeader("Origin") || "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");
  res.setHeader("Vary", "Origin");
  res.end(body);
}

module.exports = async (req, res) => {
  if (req.method === "OPTIONS") {
    res.statusCode = 204;
    res.end();
    return;
  }

  if (req.method !== "POST") {
    sendJson(res, 405, { error: "Method not allowed" });
    return;
  }

  try {
    const raw = await readBody(req);
    const body = JSON.parse(raw.toString("utf8"));
    const imageField = body.image;
    if (!imageField || typeof imageField !== "string") {
      sendJson(res, 400, { error: "Missing image (base64)" });
      return;
    }

    const threshold = Math.min(255, Math.max(0, parseInt(body.threshold, 10) || 100));
    const invert = Boolean(body.invert);

    let b64 = imageField;
    if (b64.includes(",")) b64 = b64.split(",")[1];
    const input = Buffer.from(b64, "base64");
    if (input.length > MAX_BYTES) {
      sendJson(res, 400, { error: "Image too large" });
      return;
    }

    const png = await removeBackground(input, threshold, invert);
    sendJson(res, 200, {
      png: png.toString("base64"),
      contentType: "image/png",
      engine: "node-sharp-opencv-parity",
    });
  } catch (err) {
    sendJson(res, 500, { error: err.message || "Processing failed" });
  }
};

File: api/cutout.py
"""Vercel serverless handler — runs python/background_remover.remove_background."""

from __future__ import annotations

import base64
import json
import sys
from http.server import BaseHTTPRequestHandler
from pathlib import Path

import cv2 as cv
import numpy as np

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "python"))

from background_remover import remove_background  # noqa: E402

MAX_BYTES = 12 * 1024 * 1024
DEFAULT_THRESHOLD = 100


def _cors_headers(handler: BaseHTTPRequestHandler) -> None:
    origin = handler.headers.get("Origin", "*")
    handler.send_header("Access-Control-Allow-Origin", origin)
    handler.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")
    handler.send_header("Vary", "Origin")


def _read_json(handler: BaseHTTPRequestHandler) -> dict:
    length = int(handler.headers.get("Content-Length", 0))
    if length <= 0 or length > MAX_BYTES:
        raise ValueError("Invalid request body size")
    raw = handler.rfile.read(length)
    data = json.loads(raw.decode("utf-8"))
    if not isinstance(data, dict):
        raise ValueError("Expected JSON object")
    return data


def _decode_image(payload: str) -> np.ndarray:
    if "," in payload:
        payload = payload.split(",", 1)[1]
    blob = base64.b64decode(payload, validate=True)
    if len(blob) > MAX_BYTES:
        raise ValueError("Image too large")
    arr = np.frombuffer(blob, dtype=np.uint8)
    image = cv.imdecode(arr, cv.IMREAD_COLOR)
    if image is None:
        raise ValueError("Could not decode image")
    return image


def _process(body: dict) -> bytes:
    image_b64 = body.get("image")
    if not image_b64 or not isinstance(image_b64, str):
        raise ValueError("Missing image (base64)")

    threshold = int(body.get("threshold", DEFAULT_THRESHOLD))
    invert = bool(body.get("invert", False))

    image = _decode_image(image_b64)
    result = remove_background(image, threshold, invert=invert)
    ok, buf = cv.imencode(".png", result)
    if not ok:
        raise ValueError("Could not encode PNG")
    return buf.tobytes()


class handler(BaseHTTPRequestHandler):
    def log_message(self, format: str, *args) -> None:  # noqa: A003
        return

    def do_OPTIONS(self) -> None:
        self.send_response(204)
        _cors_headers(self)
        self.end_headers()

    def do_POST(self) -> None:
        try:
            body = _read_json(self)
            png = _process(body)
            payload = json.dumps(
                {
                    "png": base64.b64encode(png).decode("ascii"),
                    "contentType": "image/png",
                }
            ).encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            _cors_headers(self)
            self.end_headers()
            self.wfile.write(payload)
        except ValueError as exc:
            self._error(400, str(exc))
        except Exception:
            self._error(500, "Processing failed")

    def _error(self, code: int, message: str) -> None:
        payload = json.dumps({"error": message}).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        _cors_headers(self)
        self.end_headers()
        self.wfile.write(payload)
